blocks/generator.go

```diff
@@ -111,16 +111,26 @@ func NewGenerator(
 
 // Start starts listening to hare output.
 func (g *Generator) Start(ctx context.Context) {
-	g.once.Do(func() {
-		ctx, g.stop = context.WithCancel(ctx)
-		g.eg.Go(func() error {
-			return g.run(ctx)
-		})
+	g.mtx.Lock()
+	defer g.mtx.Unlock()
+	if g.stop != nil {
+		return
+	}
+
+	ctx, g.stop = context.WithCancel(ctx)
+	g.eg.Go(func() error {
+		return g.run(ctx)
 	})
 }
 
 // Stop stops listening to hare output.
 func (g *Generator) Stop() {
+	g.mtx.Lock()
+	defer g.mtx.Unlock()
+
+	if g.stop == nil {
+		return
+	}
```

Comment:
I don't think `Start()` and `Stop()` need to be thread-safe. They are not meant to be executed concurrently (and they are not currently). IMHO, a nil-check would suffice here.

Reply:
It doesn't hurt 🤷, but I removed the mutex and replaced it just with the `nil` check.